src/main/java/frc/robot/scoreStates/AmpState.java

```diff
@@ -8,49 +8,81 @@
+import frc.robot.subsystems.elevator.Elevator;
+import frc.robot.subsystems.gripper.Gripper;
+import frc.robot.subsystems.gripper.GripperConstants;
 import frc.robot.swerve.SwerveDrive;
 import java.util.Set;
 import lib.Utils;
 
 public class AmpState implements ScoreState {
+    private Elevator elevator;
+    private Gripper gripper;
+    private boolean isAmpingForward = true;
+    private boolean hasTimeToTurnGripper;
+    private Translation2d ampPose;
+    private double distance;
+    private double robotRotation;
+    private Rotation2d ampRotation;
+
+    public AmpState() {
+        elevator = Elevator.getInstance();
+        gripper = Gripper.getInstance();
+    }
 
     @Override
-    public Command driveToClosestOptimalPoint() {
-        return Commands.defer(
+    public Command calculateTargets() {
+        return Commands.runOnce(
                 () -> {
-                    Translation2d ampPose;
+                    isAmpingForward = true;
                     if (isRed()) {
                         ampPose = ScoreStateConstants.AMP_POSE_RED;
                     } else {
                         ampPose = ScoreStateConstants.AMP_POSE_BLUE;
                     }
-                    Rotation2d ampRotation = ScoreStateConstants.AMP_ROTATION_NORMAL;
+                    ampRotation = ScoreStateConstants.AMP_ROTATION_NORMAL;
                     Pose2d botPose = SwerveDrive.getInstance().getBotPose();
-                    double robotRotation = botPose.getRotation().getRadians();
-                    boolean isGripperReversed =
-                            false; // TODO: replace with actual gripper.isForward
-                    double distance = Utils.getDistanceFromPoint(ampPose, botPose);
-                    boolean hasTimeToTurnGripper =
+                    robotRotation = botPose.getRotation().getRadians();
+                    distance = Utils.getDistanceFromPoint(ampPose, botPose);
+                    hasTimeToTurnGripper =
                             distance
                                     > ScoreStateConstants.MIN_DISTANCE_TO_TURN_GRIPPER.in(
                                             Units.Meters);
-
-                    if ((isGripperReversed && robotRotation < 0) || hasTimeToTurnGripper) {
+                    if ((gripper.isReversed() && robotRotation < 0) || hasTimeToTurnGripper) {
+                        isAmpingForward = false;
                         ampRotation = ScoreStateConstants.AMP_ROTATION_REVERSE;
                     }
-                    return AutoBuilder.pathfindToPose(
-                            new Pose2d(ampPose, ampRotation), Constants.AUTO_CONSTRAINTS);
-                },
-                Set.of(SwerveDrive.getInstance()));
+                });
     }
 
     @Override
-    public Command stateInitialize() {
-        return null;
+    public Command prepareSubsystems() {
+        return Commands.parallel(
+                elevator.setHeight(CommandGroupsConstants.MAX_HEIGHT),
```

Comment:
1. Perhaps another constant will be better? Real testing might show that a lower height will be optimal.
2. Can't raising the elevator be a problem if we use pathfinding and the robot needs to pass through the Stage?

Reply:
1. We'll check that in testing.
2. It can but I really don't think we'll use DTOP from far away. We may even need to put DTOP as a separate button for the driver because I think overriding the driver will throw him off.

Reply:
1. 👍🏻
2. PathPlanner knows how to switch smoothly to pathfinding, so I don't think that'll be an issue.

Reply:
I'm not saying It won't be smooth but when the driver is focusing on the match and on driving, being overridden will probably throw him off, especially if he's not the one the chooses when it happens.